File: applications/android/src/main/java/org/ducatproject/ducat/Recurring.kt
package org.ducatproject.ducat

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.util.Calendar

/**
 * Recurring bills: a schedule, not a mandate.
 *
 * What repeats here is the *asking*. Each due date mints an ordinary
 * kind-1 PaymentRequest — the same message a hand-typed request sends —
 * and §16.13 already establishes that a request carries no authority: the
 * payer approves every one on the confirm screen, or ignores it, exactly
 * as if the landlord had typed it that morning. That is why auto-sending
 * is safe where auto-*paying* never would be, and why nothing about this
 * store touches the wire, the spec, or the other phone. Cancelling is
 * local for the same reason: stopping the asking needs nobody's
 * agreement.
 */
object Recurring {
    private const val TAG = "Recurring"

    data class Bill(
        val id: String,
        val personaHex: String,
        val amountPxmr: Long,
        val note: String,
        /** Calendar-monthly when true, else every seven days. */
        val monthly: Boolean,
        /** When the next request is owed, epoch millis. */
        val nextAt: Long,
    )

    // The schedule names who pays you what, on what cadence — the shape of
    // a business relationship. Same at-rest treatment as the contacts it
    // points into.
    private fun prefs(context: Context) = securePrefs(context, "ducat_recurring")

    /**
     * Guards read-modify-write of the schedule list, like every other store
     * that rewrites its array whole ([TabStore]'s is the one this copies).
     *
     * Without it the poller and a screen wrote from lists read before each
     * other's change, and the one that matters is the *stop*: cancelling a
     * subscription while [runDue] was mid-send meant the poller's write put
     * it back, with its next date advanced, and it billed somebody again a
     * month later. The user did the one thing they could do and the app
     * undid it silently.
     */
    private val lock = Any()

    private fun <T> guarded(f: () -> T): T = synchronized(lock) { f() }

    fun all(context: Context): List<Bill> {
        val raw = prefs(context).getString("bills", null) ?: return emptyList()
        return runCatching {
            val arr = JSONArray(raw)
            (0 until arr.length()).map { i ->
                val o = arr.getJSONObject(i)
                Bill(
                    id = o.getString("id"),
                    personaHex = o.getString("who"),
                    amountPxmr = o.getLong("amt"),
                    note = o.optString("note", ""),
                    monthly = o.getBoolean("m"),
                    nextAt = o.getLong("next"),
                )
            }
        }.getOrDefault(emptyList())
    }

    private fun save(context: Context, bills: List<Bill>) {
        val arr = JSONArray()
        bills.forEach { b ->
            arr.put(
                JSONObject()
                    .put("id", b.id)
                    .put("who", b.personaHex)
                    .put("amt", b.amountPxmr)
                    .put("note", b.note)
                    .put("m", b.monthly)
                    .put("next", b.nextAt),
            )
        }
        prefs(context).edit().putString("bills", arr.toString()).apply()
        ContactStore.bump()
    }

    /** Register a schedule whose first request was just sent by hand:
     *  the next one is owed a full period from now. */
    fun add(context: Context, personaHex: String, amountPxmr: Long, note: String, monthly: Boolean) {
        val now = System.currentTimeMillis()
        val b = Bill(
            id = java.util.UUID.randomUUID().toString(),
            personaHex = personaHex,
            amountPxmr = amountPxmr,
            note = note,
            monthly = monthly,
            nextAt = advance(now, monthly),
        )
        guarded { save(context, all(context) + b) }
        DucatLog.i(TAG, "scheduled ${if (monthly) "monthly" else "weekly"} bill for ${personaHex.take(8)}")
    }

    fun stop(context: Context, id: String) =
        guarded { save(context, all(context).filterNot { it.id == id }) }

    private fun advance(from: Long, monthly: Boolean): Long =
        if (monthly) {
            // Calendar arithmetic, not day-counting: rent asked on the 31st
            // lands on the 30th where a month is short, not on the 3rd.
            Calendar.getInstance().apply {
                timeInMillis = from
                add(Calendar.MONTH, 1)
            }.timeInMillis
        } else {
            from + 7L * 24 * 60 * 60 * 1000
        }

    /**
     * The poller's hook: send what has come due.
     *
     * One request per bill per pass, advancing one period on success — so
     * a phone dark for three months surfaces three bills over three
     * passes, minutes apart, rather than skipping months that are still
     * owed or dumping them in one burst. On failure nothing advances and
     * nothing is lost: the same request is owed on the next pass.
     */
    fun runDue(context: Context) {
        val now = System.currentTimeMillis()
        val bills = all(context)
        if (bills.none { it.nextAt <= now }) return
        // Read once, not once per bill: `all()` decrypts the whole book.
        val book = ContactStore(context).all()
        val advanced = HashMap<String, Long>()
        val forgotten = HashSet<String>()
        for (b in bills) {
            if (b.nextAt > now) continue
            val c = book.firstOrNull { it.personaHex == b.personaHex }
            if (c == null) {
                // The contact was forgotten; a bill to nobody stops itself.
                DucatLog.w(TAG, "recurring bill points at a forgotten contact; dropping")
                forgotten += b.id
                continue
            }
            runCatching {
                Mailbox.send(
                    context, c,
                    b.note.ifBlank { context.getString(R.string.pay_payment_request) },
                    kind = 1, amountPxmr = b.amountPxmr,
                    payto = WalletStore(context).addressFor(b.personaHex),
                )
            }.fold(
                onSuccess = {
                    Notify.post(
                        context, c.displayName(),
                        context.getString(
                            R.string.pay_recur_notify,
                            Amounts.show(context, b.amountPxmr).primary,
                        ),
                        openChat = b.personaHex,
                    )
                    advanced[b.id] = advance(b.nextAt, b.monthly)
                },
                onFailure = { DucatLog.w(TAG, "recurring bill not sent: ${it.message}") },
            )
        }
        if (advanced.isEmpty() && forgotten.isEmpty()) return
        // **Written against the list as it stands now, not the snapshot the
        // sending started from.** A request is network-seconds, and in that
        // time somebody may have stopped a schedule or added one; writing
        // the whole snapshot back put the stopped one on the books again —
        // with its date advanced, so it billed again next period — and
        // dropped the new one. Only the schedules this pass actually acted
        // on are touched, by id, and a schedule that is no longer there is
        // simply not put back.
        guarded {
            val cur = all(context)
            val next = applyRun(cur, advanced, forgotten)
            if (next != cur) save(context, next)
        }
    }

    /**
     * The list a finished pass leaves behind, given the list as it stands
     * *now* and what the pass actually did.
     *
     * Pulled out because it is the whole of the fix and none of it needs a
     * network: a schedule stopped while its request was in flight must not
     * come back, one added meanwhile must survive, and only the ones this
     * pass really sent may move their date. The old code wrote its own
     * snapshot back over all three.
     */
    internal fun applyRun(
        current: List<Bill>,
        advanced: Map<String, Long>,
        forgotten: Set<String>,
    ): List<Bill> = current.mapNotNull { b ->
        when {
            b.id in forgotten -> null
            advanced.containsKey(b.id) -> b.copy(nextAt = advanced.getValue(b.id))
            else -> b
        }
    }
}

File: app/src/recurring.rs
//! Standing bills: a request that goes out again every week or month —
//! rent, a subscription kept by hand. The phone's `Recurring.kt`.

use serde::{Deserialize, Serialize};

use crate::contacts::{bump, now_ms};
use crate::mailbox::Outgoing;
use crate::{log, App, Error};

const TAG: &str = "Recurring";
const STORE: &str = "ducat_recurring";
pub const REQUEST_NOTE: &str = "Payment request";

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct StandingBill {
    pub id: String,
    #[serde(rename = "who")]
    pub persona_hex: String,
    #[serde(rename = "amt")]
    pub amount_pxmr: u64,
    #[serde(default)]
    pub note: String,
    #[serde(rename = "m")]
    pub monthly: bool,
    /// Millis.
    #[serde(rename = "next")]
    pub next_at: u64,
}

/// A month on, by the calendar: the same day next month, or the last
/// day of it when that month is shorter.
pub fn advance(from_ms: u64, monthly: bool) -> u64 {
    if !monthly {
        return from_ms + 7 * 24 * 60 * 60 * 1000;
    }
    let secs = from_ms / 1000;
    let tod = secs % 86_400;
    let days = (secs / 86_400) as i64;
    let (y, m, d) = civil_from_days(days);
    let (ny, nm) = if m == 12 { (y + 1, 1) } else { (y, m + 1) };
    let nd = d.min(days_in_month(ny, nm));
    let ndays = days_from_civil(ny, nm, nd);
    (ndays as u64 * 86_400 + tod) * 1000
}

fn is_leap(y: i64) -> bool {
    (y % 4 == 0 && y % 100 != 0) || y % 400 == 0
}

fn days_in_month(y: i64, m: i64) -> i64 {
    match m {
        1 | 3 | 5 | 7 | 8 | 10 | 12 => 31,
        4 | 6 | 9 | 11 => 30,
        _ => {
            if is_leap(y) {
                29
            } else {
                28
            }
        }
    }
}

fn civil_from_days(z: i64) -> (i64, i64, i64) {
    let z = z + 719_468;
    let era = z.div_euclid(146_097);
    let doe = z - era * 146_097;
    let yoe = (doe - doe / 1460 + doe / 36_524 - doe / 146_096) / 365;
    let y = yoe + era * 400;
    let doy = doe - (365 * yoe + yoe / 4 - yoe / 100);
    let mp = (5 * doy + 2) / 153;
    let d = doy - (153 * mp + 2) / 5 + 1;
    let m = if mp < 10 { mp + 3 } else { mp - 9 };
    (if m <= 2 { y + 1 } else { y }, m, d)
}

fn days_from_civil(y: i64, m: i64, d: i64) -> i64 {
    let y = if m <= 2 { y - 1 } else { y };
    let era = y.div_euclid(400);
    let yoe = y - era * 400;
    let mp = if m > 2 { m - 3 } else { m + 9 };
    let doy = (153 * mp + 2) / 5 + d - 1;
    let doe = yoe * 365 + yoe / 4 - yoe / 100 + doy;
    era * 146_097 + doe - 719_468
}

impl App {
    pub fn standing_bills(&self) -> Vec<StandingBill> {
        self.store(STORE).get("bills").unwrap_or_default()
    }

    fn save_bills(&self, bills: &[StandingBill]) -> Result<(), Error> {
        self.store(STORE).put("bills", &bills)?;
        bump();
        Ok(())
    }

    pub fn add_standing_bill(&self, persona_hex: &str, amount_pxmr: u64, note: &str, monthly: bool) -> Result<StandingBill, Error> {
        let b = StandingBill {
            id: format!("{:x}-{}", now_ms(), persona_hex.chars().take(8).collect::<String>()),
            persona_hex: persona_hex.to_string(),
            amount_pxmr,
            note: ducat_mobile::contacts::clean_display_text(note.trim().to_string()),
            monthly,
            next_at: advance(now_ms(), monthly),
        };
        let mut all = self.standing_bills();
        all.push(b.clone());
        self.save_bills(&all)?;
        log::info(TAG, format!("scheduled {} bill for {}…", if monthly { "monthly" } else { "weekly" }, &persona_hex[..8.min(persona_hex.len())]));
        Ok(b)
    }

    pub fn stop_standing_bill(&self, id: &str) -> Result<(), Error> {
        let all: Vec<StandingBill> = self.standing_bills().into_iter().filter(|b| b.id != id).collect();
        self.save_bills(&all)
    }

    /// Send every bill that has come due, and move it on; a bill whose
    /// contact is gone is dropped.
    pub fn run_due_bills(&self) {
        let now = now_ms();
        let bills = self.standing_bills();
        if !bills.iter().any(|b| b.next_at <= now) {
            return;
        }
        let mut next: Vec<StandingBill> = Vec::new();
        for b in bills {
            if b.next_at > now {
                next.push(b);
                continue;
            }
            let Some(c) = self.contact(&b.persona_hex) else {
                log::warn(TAG, "recurring bill points at a forgotten contact; dropping");
                continue;
            };
            let out = Outgoing {
                body: if b.note.trim().is_empty() { REQUEST_NOTE.into() } else { b.note.clone() },
                kind: 1,
                amount_pxmr: Some(b.amount_pxmr),
                payto: self.address_for(&b.persona_hex),
                ..Default::default()
            };
            match self.send(&c, out) {
                Ok(_) => {
                    log::info(TAG, format!("standing bill sent to {}: {} XMR", c.display_name(), crate::wallet::format_xmr(b.amount_pxmr)));
                    next.push(StandingBill { next_at: advance(b.next_at, b.monthly), ..b });
                }
                Err(e) => {
                    log::warn(TAG, format!("recurring bill not sent: {e}"));
                    next.push(b);
                }
            }
        }
        let _ = self.save_bills(&next);
    }

    /// A one-off request for payment in a thread (§16.13's kind 1).
    pub fn request_payment(&self, persona_hex: &str, amount_pxmr: u64, note: &str) -> Result<(), Error> {
        let c = self.contact(persona_hex).ok_or_else(|| Error::Refused("no such contact".into()))?;
        if amount_pxmr == 0 {
            return Err(Error::Refused("a request needs an amount".into()));
        }
        self.send(
            &c,
            Outgoing {
                body: if note.trim().is_empty() { REQUEST_NOTE.into() } else { ducat_mobile::contacts::clean_display_text(note.trim().to_string()) },
                kind: 1,
                amount_pxmr: Some(amount_pxmr),
                payto: self.address_for(persona_hex),
                ..Default::default()
            },
        )?;
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn a_month_on_is_the_same_day_or_the_last_the_month_has() {
        // 2026-01-31 12:00 UTC → 2026-02-28 12:00.
        let jan31 = (days_from_civil(2026, 1, 31) as u64 * 86_400 + 43_200) * 1000;
        let feb = advance(jan31, true);
        assert_eq!(civil_from_days((feb / 1000 / 86_400) as i64), (2026, 2, 28));
        assert_eq!((feb / 1000) % 86_400, 43_200);
        let mar15 = (days_from_civil(2026, 3, 15) as u64 * 86_400) * 1000;
        assert_eq!(civil_from_days((advance(mar15, true) / 1000 / 86_400) as i64), (2026, 4, 15));
        assert_eq!(advance(mar15, false), mar15 + 7 * 86_400 * 1000);
        assert_eq!(civil_from_days(days_from_civil(2024, 2, 29)), (2024, 2, 29));
    }
}
